### PR TITLE
[MRG] Fixing write_raw_bids to write events.tsv if they are encoded in annotations

### PR DESCRIPTION
**Pull request by adam2392**
PR Description
--------------

Closes: #581 

Lmk if you have any ways of "compressing" the `_find_events` function, which has a lot of if/else. I'm not too familiar w/ `stim`, so left that alone.

Merge checklist
---------------

Maintainer, please confirm the following before merging:

- [x] All comments resolved
- [ ] This is not your own PR
- [ ] All CIs are happy
- [x] PR title starts with [MRG]
- [x] [whats_new.rst](https://github.com/mne-tools/mne-bids/blob/master/doc/whats_new.rst) is updated
- [x] PR description includes phrase "closes <#issue-number>"

### CODE REVIEW

File: doc/whats_new.rst [hunk 1]
```diff
@@ -67,6 +67,8 @@ Bug
 - Fix :func:`mne_bids.write_raw_bids` failed BIDS validator for ``raw.info['dig'] = []`` by `Alex Rockhill`_ (`#505 <https://github.com/mne-tools/mne-bids/pull/505>`_)
 - Fix bug in :func:`mne_bids.BIDSPath.fpath` where no data file was returned when the suffix was not assigned, by `Alex Rockhill`_ (`#542 <https://github.com/mne-tools/mne-bids/pull/542>`_)
 - Ensure :func:`mne_bids.print_dir_tree` prints files and directories in alphabetical order, by `Richard Höchenberger`_ (`#563 <https://github.com/mne-tools/mne-bids/pull/563>`_)
+- Ensure :func:`mne_bids.write_raw_bids` writes events encoded in annotations of the raw file, by `Adam Li`_ (`#582 <https://github.com/mne-tools/mne-bids/pull/582>`_)
```
Comment (hoechenberger) [inline doc/whats_new.rst:70]:
```suggestion
- :func:`mne_bids.write_raw_bids` now also writes  :class:`mne.Annotations` if present in the raw file, by `Adam Li`_ (`#582 <https://github.com/mne-tools/mne-bids/pull/582>`_)
```

Comment (hoechenberger) [reply doc/whats_new.rst:70]: This is an enhancement / new feature, not a bug fix, as we never claimed to preserve Annotations. Can you move it to the respective section?

File: mne_bids/utils.py [hunk 2]
```diff
@@ -228,12 +228,18 @@ def _read_events(events_data, event_id, raw, ext, verbose=None):
                              f'found {events_data.shape[1]}')
         events = events_data
     elif 'stim' in raw:
+        print('USING STIM FOR EVENTS!!!! \n\n\n')
```
Comment (hoechenberger) [inline mne_bids/utils.py:231]: Cruft?

**Review by hoechenberger — CHANGES_REQUESTED** [review]
My view is that we should preserve **all** events here, i.e. if there are Annotations, write them; if there's `events`, write them. If we have both, write the union of both.

I also believe we should drop automated event extraction from the STIM channels, as this has proven to be buggy / not reliable for quite a number of MEG datasets in the past (at least with the default values)

**Comment by adam2392** [comment]
> My view is that we should preserve **all** events here, i.e. if there are Annotations, write them; if there's `events`, write them. If we have both, write the union of both.

What's the best way/practice for combining events? I presume we want them i) ordered and ii) have unique event IDs per different event description.

Would the best way be to simply grab events from filename/`events_data` kwarg, `stim` ch events and annotation events and then:

1. combine events_id from `events_id` kwarg and annotation events_id since IDs need to operate like a dictionary. Probably accomplish this by finding the max event_id so far and incrementing by 1 if a new event is found?
2. np.concatenate the events array and update the event IDS from step 1
3. np.argsort by onset time of event

or am I over-complicating things?

**Comment by adam2392** [comment]
There's this weird error coming up in `_read_raw_brainvision` after the changes I just pushed up... Not sure why this would be occurring tho.

**Comment by hoechenberger** [comment]
> or am I over-complicating things?

I would try to turn everything we have into Annotations, and then use `mne.events_from_annotations`. This should take care of everything, I suppose?

**Comment by adam2392** [comment]
> I would try to turn everything we have into Annotations, and then use `mne.events_from_annotations`. This should take care of everything, I suppose?
> 
> https://mne.tools/dev/auto_tutorials/intro/plot_20_events_from_raw.html#converting-between-events-arrays-and-annotations-objects

Ah okay yeah that would work. We can just append `Annotations` objects and then call `mne.events_from_annotations`.

A few questions then there because `Annotations` requires onset, duration and description:

1. For `stim` channel: `find_events` does not return events_id, which encodes the description. How should I handle that.
2. For events_data being a file, how should I construct the descriptions?
3. If `events_data` is passed in, but `event_id` is not, how should I construct the descriptions?

**Comment by agramfort** [comment]
for events we have a way as you need to pass event_id. For annotations you
have everything. So keep it simple.

>

**Comment by hoechenberger** [comment]
> 1. For `stim` channel: `find_events` does not return events_id, which encodes the description. How should I handle that.

Honestly I would drop support for this auto-event-extraction sorcery and force users to do this manually first, and then pass `events` and `event_id`.

> 2. For events_data being a file, how should I construct the descriptions?

Force users to pass an `event_id` dict.

> 3. If `events_data` is passed in, but `event_id` is not, how should I construct the descriptions?

I would raise an exception; see 2.

File: mne_bids/utils.py [hunk 3]
```diff
@@ -227,17 +226,25 @@ def _read_events(events_data, event_id, raw, ext, verbose=None):
...
-        events, event_id = events_from_annotations(raw, event_id,
-                                                   verbose=verbose)
     else:
+        events = np.empty(shape=(0, 3))
+
+    # get events from annotations
+    events_annot, event_id_annot = events_from_annotations(raw, event_id,
+                                                           verbose=verbose)
+    if event_id is not None:
+        events = np.concatenate((events, events_annot), axis=0)
```
Comment (hoechenberger) [inline mne_bids/utils.py:236]: Probably unnecessary, but could you sort based on the first column too?

```suggestion
        events = np.concatenate((events, events_annot), axis=0)
        events.sort(axis=0)
```

File: mne_bids/utils.py [hunk 4]
```diff
@@ -227,17 +226,25 @@ def _read_events(events_data, event_id, raw, ext, verbose=None):
...
+        event_id.update(event_id_annot)
+    else:
+        events = events_annot
+        event_id = event_id_annot
```
Comment (hoechenberger) [inline mne_bids/utils.py:240]: Can we turn the `if/else` block around?

```
if event_id is None:
    events = events_annot
    ...
else:
    ...
```
I always find "if not None ... else" difficult to parse, mentally ;)

File: mne_bids/utils.py [hunk 5]
```diff
@@ -14,12 +14,12 @@
 from os import path as op
 
 import numpy as np
-from mne import read_events, find_events, events_from_annotations
+from mne import (read_events, events_from_annotations)
```
Comment (hoechenberger) [inline mne_bids/utils.py:17]:
```suggestion
from mne import read_events, events_from_annotations
```

File: mne_bids/utils.py [hunk 6]
```diff
@@ -217,6 +215,7 @@ def _read_events(events_data, event_id, raw, ext, verbose=None):
         before the event or immediately after.
 
     """
+    # get events from the events_data passed in
```
Comment (hoechenberger) [inline mne_bids/utils.py:218]:
```suggestion
    # get events from events_data
```

File: mne_bids/write.py [hunk 7]
```diff
@@ -1072,7 +1087,7 @@ def write_raw_bids(raw, bids_path, events_data=None,
         logger.warning(f'Writing of electrodes.tsv is not supported '
                        f'for data type "{bids_path.datatype}". Skipping ...')
 
-    events, event_id = _read_events(events_data, event_id, raw, ext,
+    events, event_id = _read_events(events_data, event_id, raw,
                                     verbose=verbose)
```
Comment (hoechenberger) [inline mne_bids/write.py:1091]:
```suggestion
                                    verbose=False)
```

File: doc/whats_new.rst [hunk 8]
```diff
@@ -46,6 +46,7 @@ Changelog
 - Allow :func:`mne_bids.write_raw_bids` to write EEG/iEEG files from Nihon Kohden using ``mne.io.read_raw_nihon`` function, by `Adam Li`_ (`#567 <https://github.com/mne-tools/mne-bids/pull/567>`_)
 - Allow :func:`mne_bids.get_entity_vals` to ignore datatypes using ``ignore_datatypes`` kwarg, by `Adam Li`_ (`#578 <https://github.com/mne-tools/mne-bids/pull/578>`_)
 - Add ``with_key`` keyword argument to :func:`mne_bids.get_entity_vals` to allow returning the full entity strings, by `Adam Li`_ (`#578 <https://github.com/mne-tools/mne-bids/pull/578>`_)
+- :func:`mne_bids.write_raw_bids` now also writes  :class:`mne.Annotations` if present in the raw file, by `Adam Li`_ (`#582 <https://github.com/mne-tools/mne-bids/pull/582>`_)
```
Comment (agramfort) [inline doc/whats_new.rst:49]:
```suggestion
- :func:`mne_bids.write_raw_bids` now also writes :class:`mne.Annotations` if present in the raw file, by `Adam Li`_ (`#582 <https://github.com/mne-tools/mne-bids/pull/582>`_)
```

File: mne_bids/tests/test_write.py [hunk 9]
```diff
@@ -883,13 +883,37 @@ def test_eegieeg(dir_name, fname, reader, _bids_validate):
...
+
+    # read in the source raw file
     raw = reader(raw_fname)
 
+    # if there were events in the original raw dataset that
+    # were not encoded in the stim channel,
+    # they should be written even if write_raw_bids does not
+    # pass in events
+    events, events_id = mne.events_from_annotations(raw, event_id=None)
+    if events_id != {} and 'stim' not in raw:
```
Comment (agramfort) [inline mne_bids/tests/test_write.py:897]: I don't get why this `if events_id != {} and 'stim' not in raw:`

to me it should be simpler. If you have annotations present with raw it's written and you read
the written events as an annotation to make sure it's correct.

bids events.tsv is very close to an mne annotations written to disk as .txt or csv

**Comment by agramfort** [comment]
my message from this morning that did not go through:

```when you read you have annotations added. I would expect that writing back would just
take the available annotations and just work```

File: mne_bids/tests/test_write.py (hunk 9 above)
Comment (adam2392) [reply mne_bids/tests/test_write.py]: Ah okay I got rid of this `if` statement. I don't think it's necessary anymore and I took it out and the tests worked.

**Review by hoechenberger — APPROVED** [review]

File: mne_bids/utils.py [hunk 10]
```diff
@@ -227,17 +226,28 @@ def _read_events(events_data, event_id, raw, ext, verbose=None):
...
                              f'found {events_data.shape[1]}')
         events = events_data
-    elif 'stim' in raw:
-        events = find_events(raw, min_duration=0.001, initial_event=True,
-                             verbose=verbose)
-    elif ext in ['.vhdr', '.set'] and check_version('mne', '0.18'):
-        events, event_id = events_from_annotations(raw, event_id,
-                                                   verbose=verbose)
     else:
+        events = np.empty(shape=(0, 3))
```
Comment (agramfort) [inline mne_bids/utils.py:230]:
```suggestion
        events = np.empty(shape=(0, 3), dtype=int)
```

File: doc/whats_new.rst [hunk 11]
```diff
@@ -46,6 +46,7 @@ Changelog
 - Allow :func:`mne_bids.write_raw_bids` to write EEG/iEEG files from Nihon Kohden using ``mne.io.read_raw_nihon`` function, by `Adam Li`_ (`#567 <https://github.com/mne-tools/mne-bids/pull/567>`_)
 - Allow :func:`mne_bids.get_entity_vals` to ignore datatypes using ``ignore_datatypes`` kwarg, by `Adam Li`_ (`#578 <https://github.com/mne-tools/mne-bids/pull/578>`_)
 - Add ``with_key`` keyword argument to :func:`mne_bids.get_entity_vals` to allow returning the full entity strings, by `Adam Li`_ (`#578 <https://github.com/mne-tools/mne-bids/pull/578>`_)
+- :func:`mne_bids.write_raw_bids` now also writes :attr:`mne.io.Raw.annotations` to ``*_events.tsv``, by `Adam Li`_ (`#582 <https://github.com/mne-tools/mne-bids/pull/582>`_)
```
Comment (agramfort) [inline doc/whats_new.rst:49]: @hoechenberger add your name here. You did a lot.

**Review by agramfort — APPROVED** [review]
besides nitpicks LGTM !

thx @adam2392 and @hoechenberger

**Comment by hoechenberger** [comment]
Thanks @adam2392!

**Comment by adam2392** [comment]
Thanks @hoechenberger for finishing this out :)

**Pull request merged**